gnat2goto/driver/tree_walk.adb

```diff
@@ -1151,6 +1152,20 @@ package body Tree_Walk is
       Set_Identifier      (Sym, Unique_Name (E));
       Set_Type            (Sym, Symbol_Type);
 
+      if not Global_Symbol_Table.Contains (Sym_Id) then
```

Comment:
Basically, my understanding is this should only happen if we either didn't see the declaration of the type before somehow (which seems unlikely, I hope), or we couldn't parse the declaration (which would explain the many 'fixes' I suppose).

I'm still slightly skeptical this will handle all cases properly, but I'd suspect it's only going to break in cases which are otherwise broken anyway.

Reply:
I tend to agree with this analysis.  I believe it is a general Ada rule that you cannot reference an entity until it has been at least incompletely defined.